cache/redis_client.go

```diff
@@ -0,0 +1,463 @@
+	if c.MaxAsyncConcurrency <= 0 {
+		return errRedisMaxAsyncConcurrencyNotPositive
+	}
+	return nil
+}
+
+func (c *RedisClientConfig) getAddresses() ([]string, error) {
+	addresses := strings.Split(c.Endpoint, ",")
+	// Handle single configuration endpoint which resolves multiple nodes.
+	if len(addresses) == 1 {
+		host, port, err := net.SplitHostPort(addresses[0])
+		if err != nil {
+			return nil, err
+		}
+		addrs, err := net.LookupHost(host)
+		if err != nil {
+			return nil, err
+		}
+		if len(addrs) > 1 {
+			addresses = nil
+			for _, addr := range addrs {
+				addresses = append(addresses, net.JoinHostPort(addr, port))
+			}
+		}
+	}
+	return addresses, nil
+}
+
+type redisClient struct {
+	redis.UniversalClient
+	config RedisClientConfig
+
+	// Channel used to notify internal goroutines when they should quit.
+	stop chan struct{}
+
+	// Channel used to enqueue async operations.
+	asyncQueue chan func()
+
+	// getMultiGate used to enforce the max number of concurrent GetMulti() operations.
+	getMultiGate gate.Gate
+
+	// Wait group used to wait all workers on stopping.
+	workers sync.WaitGroup
+
+	logger     log.Logger
+	operations *prometheus.CounterVec
+	failures   *prometheus.CounterVec
+	skipped    *prometheus.CounterVec
+	duration   *prometheus.HistogramVec
+	dataSize   *prometheus.HistogramVec
+}
+
+// NewRedisClientWithConfig makes a new RedisClient.
+func NewRedisClientWithConfig(logger log.Logger, name string, config RedisClientConfig, reg prometheus.Registerer) (RemoteCacheClient, error) {
+	if err := config.validate(); err != nil {
+		return nil, err
+	}
+
+	addresses, err := config.getAddresses()
+	if err != nil {
+		return nil, err
+	}
+	opts := &redis.UniversalOptions{
+		Addrs:        addresses,
+		Username:     config.Username,
+		Password:     config.Password,
+		DB:           config.DB,
+		DialTimeout:  config.DialTimeout,
+		ReadTimeout:  config.ReadTimeout,
+		WriteTimeout: config.WriteTimeout,
+		MinIdleConns: config.MinIdleConns,
+		MaxConnAge:   config.MaxConnAge,
+		IdleTimeout:  config.IdleTimeout,
+	}
+
+	if config.TLSEnabled {
+		tlsConfig := config.TLSConfig
+
+		tlsClientConfig, err := newClientConfig(logger, tlsConfig.CertFile, tlsConfig.KeyFile,
+			tlsConfig.CAFile, tlsConfig.ServerName, tlsConfig.InsecureSkipVerify)
+
+		if err != nil {
+			return nil, err
+		}
+
+		opts.TLSConfig = tlsClientConfig
+	}
+
+	cli := redis.NewUniversalClient(opts)
+	if reg != nil {
+		reg = prometheus.WrapRegistererWith(prometheus.Labels{"name": name}, reg)
+	}
+
+	c := &redisClient{
+		UniversalClient: cli,
+		config:          config,
+		logger:          logger,
+		asyncQueue:      make(chan func(), config.MaxAsyncBufferSize),
+		stop:            make(chan struct{}, 1),
+		getMultiGate: gate.New(
+			prometheus.WrapRegistererWithPrefix("redis_getmulti_", reg),
+			config.MaxGetMultiConcurrency,
+		),
+	}
+
+	c.operations = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
```

Comment:
Most of these metrics are a copy of the memcached client. Could be worth moving them to a dedicated generic struct holding the cache client metrics, then used by both redis and memcached clients. The references to "redis" or "memcached" can be parametrized (pass the backend name in input to `NewCacheClientMetrics()`).